Merge common threads

scopeInfinity opened the pull request:
Issue #1950 

Current Implementation 
- Assumes child thread must be newer than parent thread
- Assign `$merged_id` as PK of root node of child thread, and set it's parent as root node of parent thread
- Marks child thread as deleted.


**Screenshots**
![screenshot from 2018-05-16 14-58-35](https://user-images.githubusercontent.com/9819066/40109328-b8042e80-591a-11e8-9c79-21656af7afe6.png)

After `Thread Y` merged into `Thread Z`
![screenshot from 2018-05-16 14-59-03](https://user-images.githubusercontent.com/9819066/40109333-bb629670-591a-11e8-85b5-7461371fc3a9.png)

scopeInfinity commented:
`Merge Threads` button will pop up list of possible parent threads.

Updated Screenshots

![screenshot from 2018-05-17 21-46-40](https://user-images.githubusercontent.com/9819066/40190859-7b0022b0-5a1d-11e8-89db-e72d2e3d9cf7.png)
![screenshot from 2018-05-17 21-48-39](https://user-images.githubusercontent.com/9819066/40190862-7cb9abb2-5a1d-11e8-86ac-7dd392fed26e.png)

andrewaikens87 commented on site/app/libraries/database/DatabaseQueries.php:
```diff
@@ -1922,6 +1922,53 @@ public function getPostsForThread($current_user, $thread_id){
+                $message = "Can't find parent thread";
+                $this->course_db->rollback();
+                return false;
+            }
```
Lines 1929-1942 has duplicate code. It might be worth creating a function that returns a boolean to whether or not a non merged thread exists.

andrewaikens87 commented:
![bad allign](https://user-images.githubusercontent.com/16356240/40253102-8031a800-5aac-11e8-899b-be779baa0fd3.png)
The popup button don't seem to stay on the same line

jchulton commented:
One potential issue is that the attachments of the child thread are not carried over after the merge. Screenshots are below. Similarly, if a thread is an announcement, it will not be after being merge if it is the child thread. I don't remember this coming up during the meetings, so we may want to talk about it during today's meeting. Other than that a potential issue is that you can't have a newer thread as the parent of an older thread, but we discussed this at our meetings.
![mergescreenshot1](https://user-images.githubusercontent.com/11095297/40314621-01ffdf50-5ce7-11e8-8d6d-099af068bae7.JPG)
![mergescreenshot2](https://user-images.githubusercontent.com/11095297/40314630-09b49d9e-5ce7-11e8-8cdc-329a044e9ca0.JPG)
![mergescreenshot3](https://user-images.githubusercontent.com/11095297/40314633-0b51113c-5ce7-11e8-9c62-496d06a78264.JPG)

andrewaikens87 commented:
You are currently storing post ids in `merged_id`, please change the `merged_id` to contain the thread_id that the row merged into.

scopeInfinity commented:
Thanks, @bmcutler @andrewaikens87 @jchulton for review.

Required changes have been made.
- Refactored code to remove duplicate code
- Fixed button alignment bug
- Fixed attachments bug on merging
- Checkbox confirming to make parent thread as announcement
  - Default selection is true if any among child or parent thread is announcement
- For now, appending child thread title to content of its first post for not losing thread title.
- Added E2E tests  #1975

andrewaikens87 commented:
There seems to be a few warnings with your tests. Take a look at [https://stackoverflow.com/questions/24450395/python-with-selenium-invalid-warning](https://stackoverflow.com/questions/24450395/python-with-selenium-invalid-warning)

`test_basic_operations_thread (e2e.test_forum.TestForumMergeThread) ... /opt/pyenv/versions/3.6.3/lib/python3.6/site-packages/selenium/webdriver/remote/webdriver.py:788: DeprecationWarning: use driver.switch_to.alert instead
  warnings.warn("use driver.switch_to.alert instead", DeprecationWarning)
ok`

`test_forum_merge_thread (e2e.test_forum.TestForumMergeThread) ... /opt/pyenv/versions/3.6.3/lib/python3.6/site-packages/selenium/webdriver/remote/webdriver.py:788: DeprecationWarning: use driver.switch_to.alert instead
  warnings.warn("use driver.switch_to.alert instead", DeprecationWarning)`

scopeInfinity commented:
Thanks @andrewaikens87 
Warnings fixed.

andrewaikens87 commented:
Just found an issue, you can recreate the screenshot error below by having two windows open. Have one merge a thread and the other stay on the child thread page and post something. It will post in the child thread a reply to a post in the parent thread.

![view](https://user-images.githubusercontent.com/16356240/40697207-2e839876-6397-11e8-9a16-4fe2fa3ccffa.png)

Merge threads

![after merge](https://user-images.githubusercontent.com/16356240/40697206-2e76fdbe-6397-11e8-9c4f-38268d6108f5.png)

Post on child thread

![after post](https://user-images.githubusercontent.com/16356240/40697205-2e647cde-6397-11e8-9352-b0d857333c64.png)

If you take a look in the database notice that the child thread 9 has a post whose parent is a part of thread 5 (which is the parent). This is from a different thread id but has the same result.

![db](https://user-images.githubusercontent.com/16356240/40696833-9b2240f6-6395-11e8-9127-54fe828c769d.png)

scopeInfinity commented:
@andrewaikens87  The bug related to publishing post on a deleted thread.
I tried fixing it by testing thread existence just before publishing post.

andrewaikens87 commented:
I think this is good. The only other thing that I think should be added is the column in the threads table to keep track of the merged thread id (currently we only keep track of the top post id of the child thread). I mentioned above for it to be added in the future but we should include it in this PR as it pertains to the code that is getting merged. @jchulton can you take another look? Thanks.

jchulton commented:
I'm still having the bug where you loose all attachments on the child branch when you merge: Screenshots below:
Child Thread:
![thread merge p1](https://user-images.githubusercontent.com/11095297/40806541-deaebe66-64ef-11e8-8043-284a1af305e4.JPG)
Parent Thread:
![thread merge p2](https://user-images.githubusercontent.com/11095297/40806545-e0a9ce22-64ef-11e8-9dd0-8a14f3e33342.JPG)
Merged Thread:
![threads merged](https://user-images.githubusercontent.com/11095297/40806546-e1edbfc8-64ef-11e8-955a-47cd2cfc4b3a.JPG)

jchulton commented:
Maybe we should change this to a WIP PR and merge it. A lot of progress has been made and this PR seems to be getting a bit long, but I don't think we should call the merging of common threads done until it works with attachments (unless we want to specifically specify that it doesn't).

andrewaikens87 commented:
@jchulton attachments seem to be working for me. I messaged you on slack to discuss more.

bmcutler commented:
Nice work :)
